externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkOperation.scala

```diff
@@ -160,16 +161,38 @@ abstract class SparkOperation(session: Session)
           case FETCH_PRIOR => iter.fetchPrior(rowSetSize);
           case FETCH_FIRST => iter.fetchAbsolute(0);
         }
-        val taken = iter.take(rowSetSize)
         resultRowSet =
-          RowSet.toTRowSet(taken.toList, resultSchema, getProtocolVersion, timeZone)
+          if (arrowEnabled) {
+            if (iter.hasNext) {
+              val taken = iter.next().asInstanceOf[Array[Byte]]
+              RowSet.toTRowSet(taken, getProtocolVersion)
+            } else {
+              RowSet.emptyTRowSet()
+            }
+          } else {
+            val taken = iter.take(rowSetSize)
+            RowSet.toTRowSet(
+              taken.toList.asInstanceOf[List[Row]],
+              resultSchema,
+              getProtocolVersion,
+              timeZone)
+          }
         resultRowSet.setStartRowOffset(iter.getPosition)
       } catch onError(cancel = true)
 
       resultRowSet
     }
 
   override def shouldRunAsync: Boolean = false
+
+  protected def arrowEnabled(): Boolean = {
+    // normalized config is required, to pass unit test
+    session.asInstanceOf[AbstractSession].normalizedConf
+      .getOrElse("kyuubi.session.result.codec", "simple")
```

Comment:
`kyuubi.operation.result.codec` because
- there are some configurations under namespace `kyuubi.operation.result.`, e.g. `kyuubi.operation.incremental.collect`, `kyuubi.operation.result.max.rows`, we should make the configuration name consistent
- instead of `connection.getResultCodec()`, the result codec can be detected from `TGetResultSetMetadataResp`, e.g. the server can append `__kyuubi_arrow_format__` to to comment of each columns in the schema to indicate that the result is encoded w/ arrow.

Reply:
let me try to append the hint to `status#infoMessages` and do the refactor

Reply:
This PR is big enough already, I think we'll rename it to `kyuubi.operation.result.codec` first and refactor it in a separate PR, what do you think?

Reply:
> rename it to `kyuubi.operation.result.codec` first ...

sgtm